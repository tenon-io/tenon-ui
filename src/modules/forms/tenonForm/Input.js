import React from 'react';

/**
 * @component
 * A default view component to use with the tenon-ui
 * smart input.
 *
 * @prop {string} contentHintText - Text to show as content
 *                hint of the smart input.
 * @prop {string} errorText - Text to show are error message
 *                of the smart input.
 * @prop required {function} getLabelProps - Prop getter function
 *              from the smart input controller for the <label>.
 * @prop required {function} getInputProps - Prop getter function
 *              from the smart input controller for the <input>.
 * @prop {function} getErrorProps - Prop getter function
 *              from the smart input controller for the
 *              error message container.
 * @prop {function} getContentHintProps - Prop getter
 *              function from the smart input controller
 *              for the content hint container.
 * @prop required {string} labelText
 * @prop {object} labelProps - Control prop collection
 *              for the <label>
 * @prop {boolean} showError - A flag to indicate when the
 *              error container should be displayed.
 *
 * NOTE: All props given to this input that does not
 * satisfy one of the above will be passed into the the
 * prop getter function of the input to enrich the input or
 * override already configured default props. Providing
 * extraneous properties that should not be rendered on an
 * <input> in the DOM can create run time errors.
 */
const Input = ({
    contentHintText,
    errorText,
    getLabelProps,
    getInputProps,
    getErrorProps,
    getContentHintProps,
    labelText,
    labelProps,
    showError,
    ...rest
}) => {
    return (
        <div className="group">
            <div className="fieldWrapper">
                <label {...getLabelProps(labelProps)}>{labelText}</label>
                <input {...getInputProps(rest)} />
            </div>
            {contentHintText && getContentHintProps ? (
                <div className="infoWrapper">
                    <span {...getContentHintProps()}>{contentHintText}</span>
                </div>
            ) : null}
            {showError && getErrorProps ? (
                <div className="errorWrapper">
                    <span {...getErrorProps()}>{errorText}</span>
                </div>
            ) : null}
        </div>
    );
};

Input.displayName = 'Input';

export default Input;
