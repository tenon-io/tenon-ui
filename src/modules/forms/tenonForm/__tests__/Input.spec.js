import React, { createRef } from 'react';
import { cleanup, render } from 'react-testing-library';
import 'jest-dom/extend-expect';
import Input from '../Input';

describe('Input', () => {
    afterEach(cleanup);

    it('should render an input and label only', () => {
        const { container } = render(
            <Input
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo' })}
                labelText="Test input"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render an error text when invalid', () => {
        const { container } = render(
            <Input
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo', 'aria-describedby': 'bar' })}
                getErrorProps={() => ({ id: 'bar' })}
                labelText="Test input"
                errorText="Test error"
                showError={true}
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a content hint if given', () => {
        const { container } = render(
            <Input
                contentHintText="Content hint test"
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo', 'aria-describedby': 'bar' })}
                getContentHintProps={() => ({ id: 'bar' })}
                labelText="Test input"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should allow for easy overriding of default props', () => {
        const { getByLabelText, getByText } = render(
            <Input
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo' })}
                labelText="Test input"
                labelProps={{ htmlFor: 'foo' }}
            />
        );

        expect(getByLabelText('Test input')).toHaveAttribute('id', 'foo');
        expect(getByText('Test input')).toHaveAttribute('for', 'foo');
    });

    it('should allow easy focus of the internal input', () => {
        const inPutRef = createRef();

        const { getByLabelText } = render(
            <Input
                ref={inPutRef}
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo' })}
                labelText="Test input"
            />
        );

        expect(getByLabelText('Test input')).toEqual(inPutRef.current);
    });

    it('should render a required text with a fallback', () => {
        const { container, rerender } = render(
            <Input
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo' })}
                labelText="Test input"
                required={true}
            />
        );

        expect(container.querySelector('span.required').innerHTML).toBe(
            '&nbsp;*'
        );

        rerender(
            <Input
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getInputProps={() => ({ id: 'foo' })}
                labelText="Test input"
                required={true}
                requiredText="required"
            />
        );

        expect(container.querySelector('span.required').innerHTML).toBe(
            '&nbsp;required'
        );
    });
});
