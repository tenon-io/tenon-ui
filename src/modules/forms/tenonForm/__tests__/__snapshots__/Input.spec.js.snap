// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render a content hint if given 1`] = `
<div
  class="group"
>
  <div
    class="fieldWrapper"
  >
    <label
      for="foo"
    >
      Test input
    </label>
    <input
      aria-describedby="bar"
      id="foo"
    />
  </div>
  <div
    class="infoWrapper"
  >
    <span
      id="bar"
    >
      Content hint test
    </span>
  </div>
</div>
`;

exports[`Input should render an error text when invalid 1`] = `
<div
  class="group"
>
  <div
    class="fieldWrapper"
  >
    <label
      for="foo"
    >
      Test input
    </label>
    <input
      aria-describedby="bar"
      id="foo"
    />
  </div>
  <div
    class="errorWrapper"
  >
    <span
      id="bar"
    >
      Test error
    </span>
  </div>
</div>
`;

exports[`Input should render an input and label only 1`] = `
<div
  class="group"
>
  <div
    class="fieldWrapper"
  >
    <label
      for="foo"
    >
      Test input
    </label>
    <input
      id="foo"
    />
  </div>
</div>
`;
